redaxo/src/core/lib/login/user_session.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+/**
+ * @package redaxo\core\login
+ */
+class rex_user_session
+{
+    use rex_singleton_trait;
+
+    private const SESSION_VAR_LAST_DB_UPDATE = 'last_db_update';
+
+    private function __construct()
+    {
+        rex_extension::register('RESPONSE_SHUTDOWN', [self::class, 'clearExpiredSessions']);
+    }
+
+    public function storeCurrentSession(): void
+    {
+        if (false === session_id()) {
+            return;
+        }
+
+        $login = new rex_backend_login();
+        $userId = $login->getSessionVar(rex_login::SESSION_IMPERSONATOR, null);
+        if (null === $userId) {
+            $userId = $login->getSessionVar(rex_login::SESSION_USER_ID);
+        }
+
+        rex_sql::factory()
+            ->setTable(rex::getTable('user_session'))
+            ->setValue('session_id', session_id())
+            ->setValue('user_id', $userId)
+            ->setValue('ip', rex_request::server('REMOTE_ADDR', 'string'))
+            ->setValue('useragent', rex_request::server('HTTP_USER_AGENT', 'string'))
+            ->setValue('starttime', rex_sql::datetime($login->getSessionVar(rex_login::SESSION_START_TIME, time())))
+            ->setValue('last_activity', rex_sql::datetime($login->getSessionVar(rex_login::SESSION_LAST_ACTIVITY)))
+            ->insertOrUpdate();
+
+        $login->setSessionVar(self::SESSION_VAR_LAST_DB_UPDATE, time());
+    }
+
+    public function clearCurrentSession(): void
+    {
+        if (false === session_id()) {
+            return;
+        }
+
+        rex_sql::factory()
+            ->setTable(rex::getTable('user_session'))
+            ->setWhere('session_id = ?', [session_id()])
+            ->delete();
+    }
+
+    public function updateLastActivity(): void
+    {
+        if (false === session_id()) {
+            return;
+        }
+
+        $login = new rex_backend_login();
+
+        // only once a minute
+        if ($login->getSessionVar(self::SESSION_VAR_LAST_DB_UPDATE, 0) > (time() - 60)) {
+            return;
+        }
+
+        $this->storeCurrentSession();
+    }
+
+    public static function clearExpiredSessions(): void
+    {
+        rex_sql::factory()
+            ->setTable(rex::getTable('user_session'))
+            ->setWhere('UNIX_TIMESTAMP(last_activity) < ?', [time() - (int) rex::getProperty('session_duration')])
+            ->delete();
```

Comment:
Hmm, überlege, ob wir vermeiden können, die Query bei jedem Backend-Aufruf aufzurufen.
Das Aufräumen ist ja nicht zwingend notwendig, oder? Also die Session werden ja auch gesperrt, auch wenn sie noch in der DB stehen.

Reicht es zum Beispiel, wenn man einfach sagt, dass man die Methode jedes Mal aufruf, wenn sich jemand frisch einloggt?

Reply:
jetzt wo du es sagst, macht es jetzt schon keinen sinn, dass der den cleanup auch macht beim shutdown im frontend :see_no_evil: 

Ja beim Login sollte IMHO auch ausreichen. Da kommen ja keine million Einträge zusammen, sodass ein zeitnaher Cleanup erforderlich.

fixe ich in #5376